File: morph/utils/logger.py
"""Colored terminal output for [morph] prefix messages."""

from __future__ import annotations
from morph.parser.errors import MorphParseError

_RESET   = "\033[0m"
_BOLD    = "\033[1m"
_DIM     = "\033[2m"
_ITALIC  = "\033[3m"

_BLACK   = "\033[30m"
_RED     = "\033[31m"
_GREEN   = "\033[32m"
_YELLOW  = "\033[33m"
_BLUE    = "\033[34m"
_MAGENTA = "\033[35m"
_CYAN    = "\033[36m"
_WHITE   = "\033[37m"

_BG_RED    = "\033[41m"
_BG_GREEN  = "\033[42m"
_BG_YELLOW = "\033[43m"
_BG_BLUE   = "\033[44m"
_BG_CYAN   = "\033[46m"


def _c(color: str, msg: str) -> str:
    return f"{color}{msg}{_RESET}"


def _wrap(prefix: str, color: str, msg: str) -> None:
    print(f"{_DIM}[{_RESET}{color}morph{_RESET}{_DIM}]{_RESET} {msg}")


def log_info(msg: str)        -> None: print(f"{_CYAN}  ◆{_RESET}  {msg}")
def log_success(msg: str)     -> None: print(f"{_GREEN}  ✓{_RESET}  {msg}")
def log_warn(msg: str)        -> None: print(f"{_YELLOW}  ⚠{_RESET}  {msg}")
def log_error(msg: str)       -> None: print(f"{_RED}  ✗{_RESET}  {msg}")

def _print_code_frame(file_path, line, col, source_lines, ctx=2) -> None:
    """Print the ┌──┐ code box with the offending line and caret."""
    loc = str(file_path) if file_path else ""
    if line:
        loc += f":{line}"
        if col:
            loc += f":{col}"

    if loc:
        print(f"  {_DIM}──>{_RESET} {_CYAN}{loc}{_RESET}")

    if source_lines and line > 0:
        start = max(0, line - 1 - ctx)
        end = min(len(source_lines), line + ctx)

        print(f"   {_DIM}──┐{_RESET}")
        for i in range(start, end):
            line_num = i + 1
            prefix = f"{_RED}{_BOLD}>{_RESET}" if line_num == line else " "
            gutter = f"{_DIM}{line_num:4}{_RESET}"
            code_line = source_lines[i].rstrip()
            print(f"   {prefix} {gutter} {_DIM}│{_RESET} {code_line}")
            if line_num == line and col:
                indent = max(0, col - 1)
                width = max(1, len(code_line) - indent)
                if width > 20:
                    width = 20
                pointer = " " * indent + _RED + "^" * width + _RESET
                print(f"     {_DIM}   │{_RESET} {pointer}")
        print(f"   {_DIM}──┘{_RESET}")


def log_parse_error(error: MorphParseError) -> None:
    """Display a parse error with file location, code context, and colors."""
    print(f"\n  {_RED}{_BOLD}error{_RESET}{_DIM}:{_RESET} {_BOLD}{str(error)}{_RESET}")
    _print_code_frame(error.file_path, error.line, error.col,
                      error.source_lines)


def log_lint_errors(errors: list) -> None:
    """Display lint findings with severity colors and code context.

    ``errors`` items are morph.checker.errors.LintError (any object with
    severity/code/message/suggestion/file_path/line/col/source_lines).
    """
    from morph.checker.errors import LintError

    errors = [e for e in errors if isinstance(e, LintError)]
    if not errors:
        return
    n_err = sum(1 for e in errors if e.severity == "error")
    n_warn = len(errors) - n_err

    summary = "lint"
    if n_err:
        summary += f" {_BOLD}{_RED}{n_err} error(s){_RESET}"
    if n_warn:
        summary += f" {_BOLD}{_YELLOW}{n_warn} warning(s){_RESET}"
    print(f"\n  {_CYAN}◆{_RESET}  {_BOLD}{summary}{_RESET}")

    for e in errors:
        color = _RED if e.severity == "error" else _YELLOW
        label = "error" if e.severity == "error" else "warning"
        print(f"\n  {color}{_BOLD}{label}{_RESET}{_DIM}:{_RESET} "
              f"{_BOLD}{e.code}{_RESET}{_DIM}:{_RESET} {e.message}")
        if e.suggestion:
            print(f"  {_DIM}  hint:{_RESET} {e.suggestion}")
        _print_code_frame(e.file_path, e.line, e.col, e.source_lines)
def log_dim(msg: str)         -> None: print(f"{_DIM}  · {msg}{_RESET}")
def log_step(msg: str)        -> None: print(f"\n{_BOLD}{_BLUE}  →{_RESET} {_BOLD}{msg}{_RESET}")
def log_header(msg: str)      -> None: print(f"\n{_BOLD}{_WHITE}{msg}{_RESET}")
def log_muted(msg: str)       -> None: print(f"{_DIM}    {msg}{_RESET}")
def log_bullet(msg: str)      -> None: print(f"{_CYAN}    •{_RESET} {msg}")
def log_key(label: str, val: str) -> None:
    print(f"    {_DIM}{label}:{_RESET} {_BOLD}{val}{_RESET}")
def log_banner(title: str) -> None:
    width = 60
    line = _DIM + "─" * width + _RESET
    print(f"\n  {line}")
    print(f"  {_BOLD}{_CYAN}  {title}{_RESET}")
    print(f"  {line}")


def print_features() -> None:
    features = [
        ("Build native UIs", "Compile .mx files → native OpenGL binary (no browser, no Electron)"),
        ("Hot Reload Dev",   "Edit code → instant window update via Unix socket IPC"),
        ("Tailwind CSS",     "500+ built-in utility classes + arbitrary values"),
        ("Custom C++ Nodes", "Drop into C++ for full OpenGL control"),
        ("Package System",   "Install community packages via `morph pkg add`"),
        ("Viewports",        "Embed raw OpenGL canvases inside your UI layout"),
    ]
    for title, desc in features:
        print(f"  {_BOLD}{_GREEN}▸{_RESET} {_BOLD}{title}{_RESET}")
        print(f"    {_DIM}{desc}{_RESET}")
    print()


def print_logo() -> None:
    logo = (
        f"  {_CYAN}{_BOLD}███╗   ███╗ ██████╗ ██████╗ ██████╗ ██╗  ██╗{_RESET}\n"
        f"  {_CYAN}{_BOLD}████╗ ████║██╔═══██╗██╔══██╗██╔══██╗██║  ██║{_RESET}\n"
        f"  {_CYAN}{_BOLD}██╔████╔██║██║   ██║██████╔╝██████╔╝███████║{_RESET}\n"
        f"  {_CYAN}{_BOLD}██║╚██╔╝██║██║   ██║██╔══██╗██╔═══╝ ██╔══██║{_RESET}\n"
        f"  {_CYAN}{_BOLD}██║ ╚═╝ ██║╚██████╔╝██║  ██║██║     ██║  ██║{_RESET}\n"
        f"  {_CYAN}{_BOLD}╚═╝     ╚═╝ ╚═════╝ ╚═╝  ╚═╝╚═╝     ╚═╝  ╚═╝{_RESET}\n"
    )
    print(logo)
    print(f"  {_DIM}Build native OpenGL Applications with HTML, CSS, and JavaScript{_RESET}")
    print(f"  {_DIM}No browser. No Electron. No WebView.{_RESET}\n")

File: crates/morphc/src/logger.rs
use colored::Colorize;

// ═══════════════════════════════════════════════════════════════
//  MORPH TUI — Better than Python version
// ═══════════════════════════════════════════════════════════════

pub fn print_logo() {
    // Animated fade-in effect (fast sequential prints)
    let lines = [
        "  ███╗   ███╗ ██████╗ ██████╗ ██████╗ ██╗  ██╗",
        "  ████╗ ████║██╔═══██╗██╔══██╗██╔══██╗██║  ██║",
        "  ██╔████╔██║██║   ██║██████╔╝██████╔╝███████║",
        "  ██║╚██╔╝██║██║   ██║██╔══██╗██╔═══╝ ██╔══██║",
        "  ██║ ╚═╝ ██║╚██████╔╝██║  ██║██║     ██║  ██║",
        "  ╚═╝     ╚═╝ ╚═════╝ ╚═╝  ╚═╝╚═╝     ╚═╝  ╚═╝",
    ];
    println!();
    for line in &lines {
        println!("{}", line.cyan().bold());
    }
    println!();
    println!("  {}  {}", "Build native OpenGL Applications with HTML, CSS, and JavaScript".white().bold(), "");
    println!("  {}", "No browser. No Electron. No WebView.".dimmed());
    println!("  {}", format!("v{} — morphc (Rust)", env!("CARGO_PKG_VERSION")).dimmed());
    println!();
}

pub fn print_features() {
    let features = [
        ("Build native UIs",  "Compile .mx files → native OpenGL binary (no browser, no Electron)"),
        ("Hot Reload Dev",    "Edit code → instant window update via Unix socket IPC"),
        ("Tailwind CSS",      "500+ built-in utility classes + arbitrary values"),
        ("Custom C++ Nodes",  "Drop into C++ for full OpenGL control"),
        ("Package System",    "Install community packages via `morph pkg add`"),
        ("Viewports",         "Embed raw OpenGL canvases inside your UI layout"),
        ("Oxc & LightningCSS","3× faster parsing, typed CSS, parallel builds"),
    ];
    for (title, desc) in features {
        println!("    {} {}", "▸".green().bold(), title.bold());
        println!("      {}", desc.dimmed());
    }
    println!();
}

// ── Core log helpers (match Python) ──

pub fn log_info(msg: &str) {
    println!("    {}  {}", "◆".cyan().bold(), msg);
}

pub fn log_success(msg: &str) {
    println!("    {}  {}", "✓".green().bold(), msg);
}

pub fn log_warn(msg: &str) {
    println!("    {}  {}", "⚠".yellow().bold(), msg);
}

pub fn log_error(msg: &str) {
    eprintln!("    {}  {}", "✗".red().bold(), msg);
}

pub fn log_step(msg: &str) {
    println!("\n  {} {}", "→".blue().bold(), msg.bold().white());
}

#[allow(dead_code)]
pub fn log_header(msg: &str) {
    println!("\n  {}", msg.bold().white());
}

pub fn log_banner(title: &str) {
    let width = 56;
    let line = "─".repeat(width);
    println!();
    println!("  {}", line.dimmed());
    println!("  {} {}", "  ◆".cyan().bold(), title.bold().white());
    println!("  {}", line.dimmed());
}

#[allow(dead_code)]
pub fn log_banner_simple(title: &str) {
    println!();
    println!("  {}", title.bold().white());
    println!();
}

pub fn log_bullet(msg: &str) {
    println!("      {} {}", "•".cyan().bold(), msg);
}

pub fn log_muted(msg: &str) {
    println!("      {}", msg.dimmed());
}

pub fn log_dim(msg: &str) {
    println!("    {} {}", "·".dimmed(), msg.dimmed());
}

pub fn log_key(label: &str, val: &str) {
    println!("      {}: {}", label.dimmed().bold(), val.bold().white());
}

pub fn log_section(msg: &str) {
    println!("\n  {}", msg.bold().white());
}

pub fn log_version(label: &str, current: &str, latest: &str) {
    let up_to_date = current == latest;
    let status = if up_to_date {
        "✓ up to date".green().bold().to_string()
    } else {
        format!("↑ {} available", "update".yellow().bold())
    };
    println!(
        "    {:<14} {} {} {}",
        label.bold(),
        format!("v{}", current).cyan(),
        format!("(latest: v{})", latest).dimmed(),
        status,
    );
}

// ── Visual dividers ──

pub fn divider() {
    println!("  {}", "─".repeat(56).dimmed());
}

pub fn divider_thick() {
    println!("  {}", "━".repeat(56).dimmed());
}

// ── Indicatif wrappers ──

pub fn spinner(msg: &str) -> indicatif::ProgressBar {
    let pb = indicatif::ProgressBar::new_spinner();
    pb.set_style(
        indicatif::ProgressStyle::default_spinner()
            .tick_strings(&["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"])
            .template("    {spinner:.cyan.bold} {msg}")
            .unwrap(),
    );
    pb.set_message(msg.to_string());
    pb.enable_steady_tick(std::time::Duration::from_millis(80));
    pb
}

#[allow(dead_code)]
pub fn progress_bar(len: u64, msg: &str) -> indicatif::ProgressBar {
    let pb = indicatif::ProgressBar::new(len);
    pb.set_style(
        indicatif::ProgressStyle::default_bar()
            .template("    {spinner:.cyan.bold} {msg} [{bar:30.cyan/blue}] {bytes}/{total_bytes} ({eta})")
            .unwrap()
            .progress_chars("█▉▊▋▌▍▎▏ "),
    );
    pb.set_message(msg.to_string());
    pb
}

// ── Code frame (matches Python's _print_code_frame) ──

pub fn print_code_frame(file_path: &str, line: usize, col: usize, source_lines: &[String]) {
    let ctx = 2usize;
    let loc = if line > 0 {
        if col > 0 { format!("{file_path}:{line}:{col}") } else { format!("{file_path}:{line}") }
    } else { file_path.to_string() };
    if !loc.is_empty() {
        println!("  {} {}", "──>".dimmed(), loc.cyan());
    }
    if source_lines.is_empty() || line == 0 { return; }
    let start = line.saturating_sub(1).saturating_sub(ctx);
    let end = (line + ctx).min(source_lines.len());
    println!("   {}","──┐".dimmed());
    for i in start..end {
        let line_num = i + 1;
        let is_target = line_num == line;
        let prefix = if is_target { ">".red().bold().to_string() } else { " ".dimmed().to_string() };
        let gutter = format!("{line_num:4}").dimmed().to_string();
        let code = source_lines[i].trim_end();
        println!("   {prefix} {gutter} {} {code}", "│".dimmed());
        if is_target && col > 0 {
            let indent = col.saturating_sub(1);
            let width = 1usize;
            let pointer = format!("{}{}", " ".repeat(indent), "^".repeat(width).red());
            println!("     {}   {} {pointer}", " ".dimmed(), "│".dimmed());
        }
    }
    println!("   {}","──┘".dimmed());
}

pub fn log_lint_errors(errors: &[morph_parser::LintError], contents: &std::collections::HashMap<String, String>) {
    if errors.is_empty() { return; }
    let n_err = errors.iter().filter(|e| e.severity == "error").count();
    let n_warn = errors.len() - n_err;
    let mut summary = "lint".to_string();
    if n_err > 0 { summary += &format!(" {}", format!("{n_err} error(s)").red().bold()); }
    if n_warn > 0 { summary += &format!(" {}", format!("{n_warn} warning(s)").yellow().bold()); }
    println!("\n  {}  {}", "◆".cyan().bold(), summary.bold());
    for e in errors {
        let _color = if e.severity == "error" { "red" } else { "yellow" };
        let label = e.severity.clone();
        let label_colored = if e.severity == "error" { label.red().bold() } else { label.yellow().bold() };
        println!("\n  {} {} {} {}", label_colored, ":".dimmed(), e.code.bold(), format!(": {}", e.message).dimmed());
        if let Some(s) = &e.suggestion {
            println!("  {}  {} {}", "hint:".dimmed(), s.dimmed(), "");
        }
        let source_lines: Vec<String> = contents.get(&e.file_path).map(|c| c.lines().map(|s| s.to_string()).collect()).unwrap_or_default();
        // Try to read file if not in contents (for cached errors)
        let source_lines = if source_lines.is_empty() {
            std::fs::read_to_string(&e.file_path).ok().map(|c| c.lines().map(|s| s.to_string()).collect::<Vec<_>>()).unwrap_or_default()
        } else { source_lines };
        print_code_frame(&e.file_path, e.line, e.col, &source_lines);
    }
}

#[allow(dead_code)]
pub fn log_parse_error(file_path: &str, line: usize, col: usize, message: &str, source_lines: &[String]) {
    println!("\n  {} {} {}", "error".red().bold(), ":".dimmed(), message.bold());
    print_code_frame(file_path, line, col, source_lines);
}

// ── Suggestion helper ──

pub fn suggest_command(input: &str) -> Option<String> {
    let commands = [
        "new", "install", "update", "dev", "build", "run",
        "check", "doctor", "cache",
    ];
    let mut best: Option<(String, f64)> = None;
    for cmd in &commands {
        let dist = strsim::levenshtein(input, cmd) as f64;
        let max_len = input.len().max(cmd.len()) as f64;
        let similarity = 1.0 - (dist / max_len);
        if similarity > 0.4 {
            if let Some((_, best_sim)) = &best {
                if similarity > *best_sim {
                    best = Some((cmd.to_string(), similarity));
                }
            } else {
                best = Some((cmd.to_string(), similarity));
            }
        }
    }
    best.map(|(cmd, _)| cmd)
}

pub fn did_you_mean(input: &str) {
    if let Some(suggestion) = suggest_command(input) {
        println!(
            "\n    {} Did you mean `{}`?",
            "?".yellow().bold(),
            suggestion.cyan().bold(),
        );
        println!(
            "    {} Run `morph {} --help` for usage.\n",
            "→".dimmed(),
            suggestion,
        );
    } else {
        println!(
            "\n    {} Unknown command `{}`. Run `morph --help` for usage.\n",
            "✗".red().bold(),
            input.red(),
        );
    }
}
